File: src/3242.design-neighbor-sum-service.3516/solution.java
class NeighborSum {
    int[][] matrix;
    int[] indexs;
    int M;
    int N;
    int[][] adjacentDeltas = new int[][]{
            new int[]{0,-1},
            new int[]{0,1},
            new int[]{-1,0},
            new int[]{1,0}
    };
    
    int[][] diagonalDeltas = new int[][]{
            new int[]{-1,-1},
            new int[]{1,1},
            new int[]{-1,1},
            new int[]{1,-1}
    };
    
    public NeighborSum(int[][] grid) {
        matrix = grid;
        M = grid.length;
        N =grid[0].length;
        indexs = new int[M*N];
        for(int i=0;i<M;i++){
            for(int j=0;j<N;j++){
                indexs[grid[i][j]] = i*N+j;
            }
        }

    }

    public int adjacentSum(int value) {
        int index = indexs[value];
        int row = index/N;
        int column = index%N;
        
        int sum = 0;
        for(int[] deltas : adjacentDeltas){
            int x = row+deltas[0];
            if(x<0 || x == M){
                continue;
            }
            int y = column+deltas[1];
            if(y<0 || y == N){
                continue;
            }
            sum += matrix[x][y];
        }
        
        return sum;
    }

    public int diagonalSum(int value) {
        int index = indexs[value];
        int row = index/N;
        int column = index%N;

        int sum = 0;
        for(int[] deltas : diagonalDeltas){
            int x = row+deltas[0];
            if(x<0 || x == M){
                continue;
            }
            int y = column+deltas[1];
            if(y<0 || y == N){
                continue;
            }
            sum += matrix[x][y];
        }

        return sum;
    }
}

/**
 * Your NeighborSum object will be instantiated and called as such:
 * NeighborSum obj = new NeighborSum(grid);
 * int param_1 = obj.adjacentSum(value);
 * int param_2 = obj.diagonalSum(value);
 */

File: src/3242.design-neighbor-sum-service.3516/solution.ts
class NeighborSum {
    private indexs:number[] = [];
    private N:number;
    private grid:number[][];
    constructor (grid: number[][]) {
        this.grid = grid;
        const N = grid.length;
        this.N = N;
        this.indexs = new Array<number>(N * N);
        for (let i = 0; i < N; i++) {
            for (let j = 0; j < N; j++) {
                this.indexs[grid[i][j]] = i * this.N + j;
            }
        }
    }

    adjacentSum (value: number): number {
        const index = this.indexs[value];
        const y = index % this.N;
        const x = (index - y) / this.N;

        let result = 0;
        const deltas = [
            [-1, 0, ],
            [1, 0, ],
            [0, -1, ],
            [0, 1, ],
        ];

        for (const [dx, dy, ] of deltas) {
            const nx = x + dx;
            if (nx < 0 || nx === this.N) {
                continue;
            }
            const ny = y + dy;
            if (ny < 0 || ny === this.N) {
                continue;
            }
            result += this.grid[nx][ny];
        }

        return result;
    }

    diagonalSum (value: number): number {
        const index = this.indexs[value];
        const y = index % this.N;
        const x = (index - y) / this.N;

        let result = 0;
        const deltas = [
            [-1, -1, ],
            [1, 1, ],
            [1, -1, ],
            [-1, 1, ],
        ];

        for (const [dx, dy, ] of deltas) {
            const nx = x + dx;
            if (nx < 0 || nx === this.N) {
                continue;
            }
            const ny = y + dy;
            if (ny < 0 || ny === this.N) {
                continue;
            }
            result += this.grid[nx][ny];
        }

        return result;
    }
}

/**
 * Your NeighborSum object will be instantiated and called as such:
 * var obj = new NeighborSum(grid)
 * var param_1 = obj.adjacentSum(value)
 * var param_2 = obj.diagonalSum(value)
 */
